utils/generate-bags/src/lib.rs

```diff
@@ -0,0 +1,277 @@
+//!
+//!    ```diff,notrust
+//!    + mod output;
+//!    - pub const BagThresholds: &'static [u64] = &[];
+//!    + pub const BagThresholds: &'static [u64] = &output::THRESHOLDS;
+//!    ```
+
+use frame_election_provider_support::VoteWeight;
+use frame_support::traits::Get;
+use std::{
+    io::Write,
+    path::{Path, PathBuf},
+};
+
+/// Compute the existential weight for the specified configuration.
+///
+/// Note that this value depends on the current issuance, a quantity known to change over time.
+/// This makes the project of computing a static value suitable for inclusion in a static,
+/// generated file _excitingly unstable_.
+fn existential_weight<T: pallet_staking::Config>(
+    total_issuance: u128,
+    minimum_balance: u128,
+) -> VoteWeight {
+    use sp_staking::currency_to_vote::CurrencyToVote;
+
+    T::CurrencyToVote::to_vote(
+        minimum_balance
+            .try_into()
+            .map_err(|_| "failed to convert minimum_balance to type Balance")
+            .unwrap(),
+        total_issuance
+            .try_into()
+            .map_err(|_| "failed to convert total_issuance to type Balance")
+            .unwrap(),
+    )
+}
+
+/// Return the path to a header file used in this repository if is exists.
+///
+/// Just searches the git working directory root for files matching certain patterns; it's
+/// pretty naive.
+fn path_to_header_file() -> Option<PathBuf> {
+    let mut workdir: &Path = &std::env::current_dir().ok()?;
+    while !workdir.join(".git").exists() {
+        workdir = workdir.parent()?;
+    }
+
+    for file_name in &["HEADER-APACHE2", "HEADER-GPL3", "HEADER", "file_header.txt"] {
+        let path = workdir.join(file_name);
+        if path.exists() {
+            return Some(path);
+        }
+    }
+    None
+}
+
+/// Create an underscore formatter: a formatter which inserts `_` every 3 digits of a number.
+fn underscore_formatter() -> num_format::CustomFormat {
+    num_format::CustomFormat::builder()
+        .grouping(num_format::Grouping::Standard)
+        .separator("_")
+        .build()
+        .expect("format described here meets all constraints")
+}
+
+/// Compute the constant ratio for the thresholds.
+///
+/// This ratio ensures that each bag, with the possible exceptions of certain small ones and the
+/// final one, is a constant multiple of the previous, while fully occupying the `VoteWeight`
+/// space.
+pub fn constant_ratio(existential_weight: VoteWeight, n_bags: usize) -> f64 {
+    ((VoteWeight::MAX as f64 / existential_weight as f64).ln() / ((n_bags - 1) as f64)).exp()
+}
+
+/// Compute the list of bag thresholds.
+///
+/// Returns a list of exactly `n_bags` elements, except in the case of overflow.
+/// The first element is always `existential_weight`.
+/// The last element is always `VoteWeight::MAX`.
+///
+/// All other elements are computed from the previous according to the formula
+/// `threshold[k + 1] = (threshold[k] * ratio).max(threshold[k] + 1);`
```

Comment:
I assume yes, but please double check with John if this way of computing the bags thresholds is fine.

Reply:
asked John, waiting on reply

Reply:
this is fine, confirmed